// mongodb/lib/operations/execute_operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeOperation = executeOperation;
exports.autoConnect = autoConnect;
const promises_1 = require("timers/promises");
const constants_1 = require("../cmap/wire_protocol/constants");
const error_1 = require("../error");
const read_preference_1 = require("../read_preference");
const common_1 = require("../sdam/common");
const server_selection_1 = require("../sdam/server_selection");
const timeout_1 = require("../timeout");
const utils_1 = require("../utils");
const aggregate_1 = require("./aggregate");
const operation_1 = require("./operation");
const run_command_1 = require("./run_command");
const MMAPv1_RETRY_WRITES_ERROR_CODE = error_1.MONGODB_ERROR_CODES.IllegalOperation;
const MMAPv1_RETRY_WRITES_ERROR_MESSAGE = 'This MongoDB deployment does not support retryable writes. Please add retryWrites=false to your connection string.';

async function executeOperation(client, operation, timeoutContext) {
    if (!(operation instanceof operation_1.AbstractOperation)) {
        
        throw new error_1.MongoRuntimeError('This method requires a valid operation instance');
    }
    const topology = client.topology == null
        ? await (0, utils_1.abortable)(autoConnect(client), operation.options)
        : client.topology;
    
    
    let session = operation.session;
    let owner;
    if (session == null) {
        owner = Symbol();
        session = client.startSession({ owner, explicit: false });
    }
    else if (session.hasEnded) {
        throw new error_1.MongoExpiredSessionError('Use of expired sessions is not permitted');
    }
    else if (session.snapshotEnabled &&
        (0, utils_1.maxWireVersion)(topology) < constants_1.MIN_SUPPORTED_SNAPSHOT_READS_WIRE_VERSION) {
        throw new error_1.MongoCompatibilityError('Snapshot reads require MongoDB 5.0 or later');
    }
    else if (session.client !== client) {
        throw new error_1.MongoInvalidArgumentError('ClientSession must be from the same MongoClient');
    }
    operation.session ??= session;
    const readPreference = operation.readPreference ?? read_preference_1.ReadPreference.primary;
    const inTransaction = !!session?.inTransaction();
    const hasReadAspect = operation.hasAspect(operation_1.Aspect.READ_OPERATION);
    if (inTransaction &&
        !readPreference.equals(read_preference_1.ReadPreference.primary) &&
        (hasReadAspect || operation.commandName === 'runCommand')) {
        throw new error_1.MongoTransactionError(`Read preference in a transaction must be primary, not: ${readPreference.mode}`);
    }
    if (session?.isPinned && session.transaction.isCommitted && !operation.bypassPinningCheck) {
        session.unpin();
    }
    timeoutContext ??= timeout_1.TimeoutContext.create({
        session,
        serverSelectionTimeoutMS: client.s.options.serverSelectionTimeoutMS,
        waitQueueTimeoutMS: client.s.options.waitQueueTimeoutMS,
        timeoutMS: operation.options.timeoutMS
    });
    try {
        return await executeOperationWithRetries(operation, {
            topology,
            timeoutContext,
            session,
            readPreference
        });
    }
    finally {
        if (session?.owner != null && session.owner === owner) {
            await session.endSession();
        }
    }
}

async function autoConnect(client) {
    if (client.topology == null) {
        if (client.s.hasBeenClosed) {
            throw new error_1.MongoNotConnectedError('Client must be connected before running operations');
        }
        client.s.options.__skipPingOnConnect = true;
        try {
            await client.connect();
            if (client.topology == null) {
                throw new error_1.MongoRuntimeError('client.connect did not create a topology but also did not throw');
            }
            return client.topology;
        }
        finally {
            delete client.s.options.__skipPingOnConnect;
        }
    }
    return client.topology;
}

const BASE_BACKOFF_MS = 100;

const MAX_BACKOFF_MS = 10_000;

async function executeOperationWithRetries(operation, { topology, timeoutContext, session, readPreference }) {
    let selector;
    if (operation.hasAspect(operation_1.Aspect.MUST_SELECT_SAME_SERVER)) {
        
        
        
        selector = (0, server_selection_1.sameServerSelector)(operation.server?.description);
    }
    else if (operation instanceof aggregate_1.AggregateOperation && operation.hasWriteStage) {
        
        
        selector = (0, server_selection_1.secondaryWritableServerSelector)(topology.commonWireVersion, readPreference);
    }
    else {
        selector = readPreference;
    }
    let server = await topology.selectServer(selector, {
        session,
        operationName: operation.commandName,
        timeoutContext,
        signal: operation.options.signal,
        deprioritizedServers: new server_selection_1.DeprioritizedServers()
    });
    const hasReadAspect = operation.hasAspect(operation_1.Aspect.READ_OPERATION);
    const hasWriteAspect = operation.hasAspect(operation_1.Aspect.WRITE_OPERATION);
    const inTransaction = session?.inTransaction() ?? false;
    const willRetryRead = topology.s.options.retryReads && !inTransaction && operation.canRetryRead;
    const willRetryWrite = topology.s.options.retryWrites &&
        !inTransaction &&
        (0, utils_1.supportsRetryableWrites)(server) &&
        operation.canRetryWrite;
    const willRetry = operation.hasAspect(operation_1.Aspect.RETRYABLE) &&
        session != null &&
        ((hasReadAspect && willRetryRead) || (hasWriteAspect && willRetryWrite));
    if (hasWriteAspect && willRetryWrite && session != null) {
        operation.options.willRetryWrite = true;
        session.incrementTransactionNumber();
    }
    const deprioritizedServers = new server_selection_1.DeprioritizedServers();
    let maxAttempts = typeof operation.maxAttempts === 'number'
        ? operation.maxAttempts
        : willRetry
            ? timeoutContext.csotEnabled()
                ? Infinity
                : 2
            : 1;
    let error = null;
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
        operation.attemptsMade = attempt + 1;
        operation.server = server;
        try {
            try {
                const result = await server.command(operation, timeoutContext);
                return operation.handleOk(result);
            }
            catch (error) {
                return operation.handleError(error);
            }
        }
        catch (operationError) {
            
            if (!(operationError instanceof error_1.MongoError))
                throw operationError;
            
            
            if (error == null) {
                error = operationError;
            }
            else {
                if (!operationError.hasErrorLabel(error_1.MongoErrorLabel.NoWritesPerformed)) {
                    error = operationError;
                }
            }
            
            timeoutContext.clear();
            if (hasWriteAspect && operationError.code === MMAPv1_RETRY_WRITES_ERROR_CODE) {
                throw new error_1.MongoServerError({
                    message: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
                    errmsg: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
                    originalError: operationError
                });
            }
            if (!canRetry(operation, operationError)) {
                throw error;
            }
            if (operationError.hasErrorLabel(error_1.MongoErrorLabel.SystemOverloadedError)) {
                const maxOverloadAttempts = topology.s.options.maxAdaptiveRetries + 1;
                maxAttempts = Math.min(maxOverloadAttempts, operation.maxAttempts ?? maxOverloadAttempts);
            }
            if (attempt + 1 >= maxAttempts) {
                throw error;
            }
            if (operationError instanceof error_1.MongoNetworkError &&
                operation.hasAspect(operation_1.Aspect.CURSOR_CREATING) &&
                session != null &&
                session.isPinned &&
                !session.inTransaction()) {
                session.unpin({ force: true, forceClear: true });
            }
            if (operationError.hasErrorLabel(error_1.MongoErrorLabel.SystemOverloadedError) &&
                operation.hasAspect(operation_1.Aspect.CURSOR_CREATING) &&
                session != null &&
                session.isPinned &&
                !session.inTransaction()) {
                session.unpin({ force: true });
            }
            if (operationError.hasErrorLabel(error_1.MongoErrorLabel.SystemOverloadedError)) {
                const backoffMS = Math.random() * Math.min(MAX_BACKOFF_MS, BASE_BACKOFF_MS * 2 ** attempt);
                
                if (timeoutContext.csotEnabled() && backoffMS > timeoutContext.remainingTimeMS) {
                    throw error;
                }
                await (0, promises_1.setTimeout)(backoffMS);
            }
            if (topology.description.type === common_1.TopologyType.Sharded ||
                (operationError.hasErrorLabel(error_1.MongoErrorLabel.SystemOverloadedError) &&
                    topology.s.options.enableOverloadRetargeting)) {
                deprioritizedServers.add(server.description);
            }
            server = await topology.selectServer(selector, {
                session,
                operationName: operation.commandName,
                deprioritizedServers,
                signal: operation.options.signal
            });
            if (hasWriteAspect &&
                !(0, utils_1.supportsRetryableWrites)(server) &&
                !operationError.hasErrorLabel(error_1.MongoErrorLabel.SystemOverloadedError)) {
                throw new error_1.MongoUnexpectedServerResponseError('Selected server does not support retryable writes');
            }
            
            
            if (operation.hasAspect(operation_1.Aspect.COMMAND_BATCHING)) {
                operation.resetBatch();
            }
        }
    }
    throw (error ??
        new error_1.MongoRuntimeError('Should never happen: operation execution loop terminated but no error was recorded.'));
    function canRetry(operation, error) {
        
        
        
        
        if (error.hasErrorLabel(error_1.MongoErrorLabel.SystemOverloadedError) &&
            error.hasErrorLabel(error_1.MongoErrorLabel.RetryableError)) {
            
            if (operation instanceof run_command_1.RunCommandOperation) {
                return topology.s.options.retryReads && topology.s.options.retryWrites;
            }
            
            if (operation instanceof aggregate_1.AggregateOperation && operation.hasWriteStage) {
                return topology.s.options.retryWrites;
            }
            
            const canRetryAsRead = hasReadAspect && topology.s.options.retryReads;
            const canRetryAsWrite = hasWriteAspect && topology.s.options.retryWrites;
            return canRetryAsRead || canRetryAsWrite;
        }
        
        if (operation instanceof run_command_1.RunCommandOperation) {
            return false;
        }
        
        if (operation.hasAspect(operation_1.Aspect.COMMAND_BATCHING)) {
            return operation.canRetryWrite && (0, error_1.isRetryableWriteError)(error);
        }
        return ((hasWriteAspect && willRetryWrite && (0, error_1.isRetryableWriteError)(error)) ||
            (hasReadAspect && willRetryRead && (0, error_1.isRetryableReadError)(error)));
    }
}
