// mongoose/lib/helpers/model/applyHooks.js
'use strict';

const { buildMiddlewareFilter } = require('../buildMiddlewareFilter');



module.exports = applyHooks;



applyHooks.middlewareFunctions = [
  'deleteOne',
  'remove',
  'save',
  'updateOne',
  'validate',
  'init'
];



const alreadyHookedFunctions = new Set(applyHooks.middlewareFunctions.flatMap(fn => ([fn, `$__${fn}`])));



function applyHooks(model, schema, options) {
  options = options || {};

  const kareemOptions = {
    useErrorHandlers: true,
    numCallbackParams: 1,
    nullResultByDefault: true,
    contextParameter: true
  };
  const objToDecorate = options.decorateDoc ? model : model.prototype;
  model.$appliedHooks = true;
  for (const key of Object.keys(schema.paths)) {
    let type = schema.paths[key];
    let childModel = null;

    const result = findChildModel(type);
    if (result) {
      childModel = result.childModel;
      type = result.type;
    } else {
      continue;
    }

    if (childModel.$appliedHooks) {
      continue;
    }

    applyHooks(childModel, type.schema, {
      ...options,
      decorateDoc: false,
      isChildSchema: true
    });
    if (childModel.discriminators != null) {
      const keys = Object.keys(childModel.discriminators);
      for (const key of keys) {
        applyHooks(childModel.discriminators[key],
          childModel.discriminators[key].schema, options);
      }
    }
  }

  
  
  

  const middleware = schema._getDocumentMiddleware();

  model._middleware = middleware;

  objToDecorate.$__init = middleware.
    createWrapperSync('init', objToDecorate.$__init, null, {
      ...kareemOptions,
      getOptions: (args) => {
        const opts = args[1];
        return {
          pre: { filter: buildMiddlewareFilter(opts, 'pre') },
          post: { filter: buildMiddlewareFilter(opts, 'post') }
        };
      }
    });

  
  const customMethods = Object.keys(schema.methods);
  const customMethodOptions = Object.assign({}, kareemOptions, {
    
    
    
    
    checkForPromise: true
  });
  for (const method of customMethods) {
    if (alreadyHookedFunctions.has(method)) {
      continue;
    }
    if (!middleware.hasHooks(method)) {
      
      
      
      continue;
    }
    const originalMethod = objToDecorate[method];
    objToDecorate[`$__${method}`] = objToDecorate[method];
    objToDecorate[method] = middleware.
      createWrapper(method, originalMethod, null, customMethodOptions);
  }
}



function findChildModel(curType) {
  if (curType.$isSingleNested || curType.$isMongooseDocumentArray) {
    return { childModel: curType.Constructor, type: curType };
  }
  if (curType.instance === 'Array') {
    const embedded = curType.getEmbeddedSchemaType();
    if (embedded) {
      return findChildModel(embedded);
    }
  }
  if (curType.instance === 'Map') {
    const mapType = curType.getEmbeddedSchemaType();
    if (mapType) {
      return findChildModel(mapType);
    }
  }
  return null;
}
