// mongodb/lib/cmap/wire_protocol/on_demand/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OnDemandDocument = void 0;
const bson_1 = require("../../../bson");
const BSONElementOffset = {
    type: 0,
    nameOffset: 1,
    nameLength: 2,
    offset: 3,
    length: 4
};

class OnDemandDocument {
    constructor(bson, offset = 0, isArray = false, 
    
    elements) {
        
        this.cache = Object.create(null);
        
        this.indexFound = Object.create(null);
        this.bson = bson;
        this.offset = offset;
        this.isArray = isArray;
        this.elements = elements ?? (0, bson_1.parseToElementsToArray)(this.bson, offset);
    }
    
    isElementName(name, element) {
        const nameLength = element[BSONElementOffset.nameLength];
        const nameOffset = element[BSONElementOffset.nameOffset];
        if (name.length !== nameLength)
            return false;
        const nameEnd = nameOffset + nameLength;
        for (let byteIndex = nameOffset, charIndex = 0; charIndex < name.length && byteIndex < nameEnd; charIndex++, byteIndex++) {
            if (this.bson[byteIndex] !== name.charCodeAt(charIndex))
                return false;
        }
        return true;
    }
    
    getElement(name) {
        const cachedElement = this.cache[name];
        if (cachedElement === false)
            return null;
        if (cachedElement != null) {
            return cachedElement;
        }
        if (typeof name === 'number') {
            if (this.isArray) {
                if (name < this.elements.length) {
                    const element = this.elements[name];
                    const cachedElement = { element, value: undefined };
                    this.cache[name] = cachedElement;
                    this.indexFound[name] = true;
                    return cachedElement;
                }
                else {
                    return null;
                }
            }
            else {
                return null;
            }
        }
        for (let index = 0; index < this.elements.length; index++) {
            const element = this.elements[index];
            
            if (!(index in this.indexFound) && this.isElementName(name, element)) {
                const cachedElement = { element, value: undefined };
                this.cache[name] = cachedElement;
                this.indexFound[index] = true;
                return cachedElement;
            }
        }
        this.cache[name] = false;
        return null;
    }
    toJSValue(element, as) {
        const type = element[BSONElementOffset.type];
        const offset = element[BSONElementOffset.offset];
        const length = element[BSONElementOffset.length];
        if (as !== type) {
            return null;
        }
        switch (as) {
            case bson_1.BSONType.null:
            case bson_1.BSONType.undefined:
                return null;
            case bson_1.BSONType.double:
                return bson_1.NumberUtils.getFloat64LE(this.bson, offset);
            case bson_1.BSONType.int:
                return bson_1.NumberUtils.getInt32LE(this.bson, offset);
            case bson_1.BSONType.long:
                return bson_1.NumberUtils.getBigInt64LE(this.bson, offset);
            case bson_1.BSONType.bool:
                return Boolean(this.bson[offset]);
            case bson_1.BSONType.objectId:
                return new bson_1.ObjectId(this.bson.subarray(offset, offset + 12));
            case bson_1.BSONType.timestamp:
                return new bson_1.Timestamp(bson_1.NumberUtils.getBigInt64LE(this.bson, offset));
            case bson_1.BSONType.string:
                return bson_1.ByteUtils.toUTF8(this.bson, offset + 4, offset + length - 1, false);
            case bson_1.BSONType.binData: {
                const totalBinarySize = bson_1.NumberUtils.getInt32LE(this.bson, offset);
                const subType = this.bson[offset + 4];
                if (subType === 2) {
                    const subType2BinarySize = bson_1.NumberUtils.getInt32LE(this.bson, offset + 1 + 4);
                    if (subType2BinarySize < 0)
                        throw new bson_1.BSONError('Negative binary type element size found for subtype 0x02');
                    if (subType2BinarySize > totalBinarySize - 4)
                        throw new bson_1.BSONError('Binary type with subtype 0x02 contains too long binary size');
                    if (subType2BinarySize < totalBinarySize - 4)
                        throw new bson_1.BSONError('Binary type with subtype 0x02 contains too short binary size');
                    return new bson_1.Binary(this.bson.subarray(offset + 1 + 4 + 4, offset + 1 + 4 + 4 + subType2BinarySize), 2);
                }
                return new bson_1.Binary(this.bson.subarray(offset + 1 + 4, offset + 1 + 4 + totalBinarySize), subType);
            }
            case bson_1.BSONType.date:
                
                return new Date(Number(bson_1.NumberUtils.getBigInt64LE(this.bson, offset)));
            case bson_1.BSONType.object:
                return new OnDemandDocument(this.bson, offset);
            case bson_1.BSONType.array:
                return new OnDemandDocument(this.bson, offset, true);
            default:
                throw new bson_1.BSONError(`Unsupported BSON type: ${as}`);
        }
    }
    
    size() {
        return this.elements.length;
    }
    
    has(name) {
        const cachedElement = this.cache[name];
        if (cachedElement === false)
            return false;
        if (cachedElement != null)
            return true;
        return this.getElement(name) != null;
    }
    get(name, as, required) {
        const element = this.getElement(name);
        if (element == null) {
            if (required === true) {
                throw new bson_1.BSONError(`BSON element "${name}" is missing`);
            }
            else {
                return null;
            }
        }
        if (element.value == null) {
            const value = this.toJSValue(element.element, as);
            if (value == null) {
                if (required === true) {
                    throw new bson_1.BSONError(`BSON element "${name}" is missing`);
                }
                else {
                    return null;
                }
            }
            
            element.value = value;
        }
        return element.value;
    }
    getNumber(name, required) {
        const maybeBool = this.get(name, bson_1.BSONType.bool);
        const bool = maybeBool == null ? null : maybeBool ? 1 : 0;
        const maybeLong = this.get(name, bson_1.BSONType.long);
        const long = maybeLong == null ? null : Number(maybeLong);
        const result = bool ?? long ?? this.get(name, bson_1.BSONType.int) ?? this.get(name, bson_1.BSONType.double);
        if (required === true && result == null) {
            throw new bson_1.BSONError(`BSON element "${name}" is missing`);
        }
        return result;
    }
    
    toObject(options) {
        return (0, bson_1.deserialize)(this.bson, {
            ...options,
            index: this.offset,
            allowObjectSmallerThanBufferSize: true
        });
    }
    
    toBytes() {
        const size = bson_1.NumberUtils.getInt32LE(this.bson, this.offset);
        return this.bson.subarray(this.offset, this.offset + size);
    }
}
exports.OnDemandDocument = OnDemandDocument;
