// mongoose/lib/error/setOptionError.js


'use strict';

const MongooseError = require('./mongooseError');
const util = require('util');
const combinePathErrors = require('../helpers/error/combinePathErrors');



class SetOptionError extends MongooseError {

  constructor() {
    super('');

    this.errors = {};
  }

  
  toString() {
    return combinePathErrors(this);
  }

  
  inspect() {
    return Object.assign(new Error(this.message), this);
  }

  
  addError(key, error) {
    if (error instanceof SetOptionError) {
      const { errors } = error;
      for (const optionKey of Object.keys(errors)) {
        this.addError(optionKey, errors[optionKey]);
      }

      return;
    }

    this.errors[key] = error;
    this.message = combinePathErrors(this);
  }
}


if (util.inspect.custom) {
  
  SetOptionError.prototype[util.inspect.custom] = SetOptionError.prototype.inspect;
}


Object.defineProperty(SetOptionError.prototype, 'toJSON', {
  enumerable: false,
  writable: false,
  configurable: true,
  value: function() {
    return Object.assign({}, this, { name: this.name, message: this.message });
  }
});


Object.defineProperty(SetOptionError.prototype, 'name', {
  value: 'SetOptionError'
});

class SetOptionInnerError extends MongooseError {
  
  constructor(key) {
    super(`"${key}" is not a valid option to set`);
  }
}

SetOptionError.SetOptionInnerError = SetOptionInnerError;



module.exports = SetOptionError;
