// mongoose/lib/schemaType.js
'use strict';



const MongooseError = require('./error/index');
const SchemaTypeOptions = require('./options/schemaTypeOptions');
const $exists = require('./schema/operators/exists');
const $type = require('./schema/operators/type');
const clone = require('./helpers/clone');
const handleImmutable = require('./helpers/schematype/handleImmutable');
const isAsyncFunction = require('./helpers/isAsyncFunction');
const isSimpleValidator = require('./helpers/isSimpleValidator');
const schemaTypeSymbol = require('./helpers/symbols').schemaTypeSymbol;
const utils = require('./utils');
const validatorErrorSymbol = require('./helpers/symbols').validatorErrorSymbol;
const documentIsModified = require('./helpers/symbols').documentIsModified;

const populateModelSymbol = require('./helpers/symbols').populateModelSymbol;

const CastError = MongooseError.CastError;
const ValidatorError = MongooseError.ValidatorError;

const setOptionsForDefaults = { _skipMarkModified: true };



function SchemaType(path, options, instance, parentSchema) {
  this[schemaTypeSymbol] = true;
  this.path = path;
  this.instance = instance;
  this.schemaName = this.constructor.schemaName;
  this.validators = [];
  this.getters = Object.hasOwn(this.constructor, 'getters') ?
    this.constructor.getters.slice() :
    [];
  this.setters = Object.hasOwn(this.constructor, 'setters') ?
    this.constructor.setters.slice() :
    [];

  this.splitPath();

  options = options || {};
  const defaultOptions = this.constructor.defaultOptions || {};
  const defaultOptionsKeys = Object.keys(defaultOptions);

  for (const option of defaultOptionsKeys) {
    if (option === 'validate') {
      this.validate(defaultOptions.validate);
    } else if (Object.hasOwn(defaultOptions, option) && !Object.hasOwn(options, option)) {
      options[option] = defaultOptions[option];
    }
  }

  if (options.select == null) {
    delete options.select;
  }

  const Options = this.OptionsConstructor || SchemaTypeOptions;
  this.options = new Options(options);
  this.parentSchema = parentSchema;
  this._index = null;

  if (utils.hasUserDefinedProperty(this.options, 'immutable')) {
    this.$immutable = this.options.immutable;

    handleImmutable(this);
  }

  const keys = Object.keys(this.options);
  for (const prop of keys) {
    if (prop === 'cast') {
      if (Array.isArray(this.options[prop])) {
        this.castFunction.apply(this, this.options[prop]);
      } else {
        this.castFunction(this.options[prop]);
      }
      continue;
    }
    if (utils.hasUserDefinedProperty(this.options, prop) && typeof this[prop] === 'function') {
      
      if (prop === 'index' && this._index) {
        if (options.index === false) {
          const index = this._index;
          if (typeof index === 'object' && index != null) {
            if (index.unique) {
              throw new MongooseError('Path "' + this.path + '" may not have `index` ' +
                'set to false and `unique` set to true');
            }
            if (index.sparse) {
              throw new MongooseError('Path "' + this.path + '" may not have `index` ' +
                'set to false and `sparse` set to true');
            }
          }

          this._index = false;
        }
        continue;
      }

      const val = options[prop];
      
      if (prop === 'default') {
        this.default(val);
        continue;
      }

      const opts = Array.isArray(val) ? val : [val];

      this[prop].apply(this, opts);
    }
  }

  Object.defineProperty(this, '$$context', {
    enumerable: false,
    configurable: false,
    writable: true,
    value: null
  });
}



SchemaType.prototype.OptionsConstructor = SchemaTypeOptions;



SchemaType.prototype.path;



SchemaType.prototype.toJSON = function toJSON() {
  const res = { ...this };
  delete res.parentSchema;
  return res;
};



SchemaType.prototype._createJSONSchemaTypeDefinition = function _createJSONSchemaTypeDefinition(type, bsonType, options) {
  const useBsonType = options?.useBsonType;
  const isRequired = options?._overrideRequired ??
    (this.options.required == null ?
      (options?._defaultRequired === true || this.path === '_id') :
      this.options.required && typeof this.options.required !== 'function');
  const allowNull = this.options.allowNull !== false;

  if (useBsonType) {
    if (isRequired || !allowNull) {
      return { bsonType };
    }
    return { bsonType: [bsonType, 'null'] };
  }

  if (isRequired || !allowNull) {
    return { type };
  }
  return { type: [type, 'null'] };
};



SchemaType.prototype.validators;



SchemaType.prototype.isRequired;



SchemaType.prototype.splitPath = function() {
  if (this._presplitPath != null) {
    return this._presplitPath;
  }
  if (this.path == null) {
    return undefined;
  }

  this._presplitPath = this.path.indexOf('.') === -1 ? [this.path] : this.path.split('.');
  return this._presplitPath;
};



SchemaType.cast = function cast(caster) {
  if (arguments.length === 0) {
    return this._cast;
  }
  if (caster === false) {
    caster = v => v;
  }
  this._cast = caster;

  return this._cast;
};



SchemaType.prototype.castFunction = function castFunction(caster, message) {
  if (arguments.length === 0) {
    return this._castFunction;
  }

  if (caster === false) {
    caster = this.constructor._defaultCaster || (v => v);
  }
  if (typeof caster === 'string') {
    this._castErrorMessage = caster;
    return this._castFunction;
  }
  if (caster != null) {
    this._castFunction = caster;
  }
  if (message != null) {
    this._castErrorMessage = message;
  }

  return this._castFunction;
};



SchemaType.prototype.cast = function cast() {
  throw new MongooseError('Base SchemaType class does not implement a `cast()` function');
};



SchemaType.set = function set(option, value) {
  if (!Object.hasOwn(this, 'defaultOptions')) {
    this.defaultOptions = Object.assign({}, this.defaultOptions);
  }
  this.defaultOptions[option] = value;
};



SchemaType.get = function(getter) {
  this.getters = Object.hasOwn(this, 'getters') ? this.getters : [];
  this.getters.push(getter);
};



SchemaType.prototype.default = function(val) {
  if (arguments.length === 1) {
    if (val === void 0) {
      this.defaultValue = void 0;
      return void 0;
    }

    if (val?.instanceOfSchema) {
      throw new MongooseError('Cannot set default value of path `' + this.path +
        '` to a mongoose Schema instance.');
    }

    this.defaultValue = val;
    return this.defaultValue;
  } else if (arguments.length > 1) {
    this.defaultValue = [...arguments];
  }
  return this.defaultValue;
};



SchemaType.prototype.index = function(options) {
  this._index = options;
  utils.expires(this._index);
  return this;
};



SchemaType.prototype.unique = function unique(value, message) {
  if (this._index === false) {
    if (!value) {
      return;
    }
    throw new MongooseError('Path "' + this.path + '" may not have `index` set to ' +
      'false and `unique` set to true');
  }

  if (!Object.hasOwn(this.options, 'index') && value === false) {
    return this;
  }

  if (this._index == null || this._index === true) {
    this._index = {};
  } else if (typeof this._index === 'string') {
    this._index = { type: this._index };
  }

  this._index.unique = !!value;
  if (typeof message === 'string') {
    this._duplicateKeyErrorMessage = message;
  }
  return this;
};



SchemaType.prototype.text = function(bool) {
  if (this._index === false) {
    if (!bool) {
      return this;
    }
    throw new MongooseError('Path "' + this.path + '" may not have `index` set to ' +
      'false and `text` set to true');
  }

  if (!Object.hasOwn(this.options, 'index') && bool === false) {
    return this;
  }

  if (this._index === null || this._index === undefined ||
    typeof this._index === 'boolean') {
    this._index = {};
  } else if (typeof this._index === 'string') {
    this._index = { type: this._index };
  }

  this._index.text = bool;
  return this;
};



SchemaType.prototype.sparse = function(bool) {
  if (this._index === false) {
    if (!bool) {
      return this;
    }
    throw new MongooseError('Path "' + this.path + '" may not have `index` set to ' +
      'false and `sparse` set to true');
  }

  if (!Object.hasOwn(this.options, 'index') && bool === false) {
    return this;
  }

  if (this._index == null || typeof this._index === 'boolean') {
    this._index = {};
  } else if (typeof this._index === 'string') {
    this._index = { type: this._index };
  }

  this._index.sparse = bool;
  return this;
};



SchemaType.prototype.immutable = function(bool) {
  this.$immutable = bool;
  handleImmutable(this);

  return this;
};



SchemaType.prototype.transform = function(fn) {
  this.options.transform = fn;

  return this;
};



SchemaType.prototype.set = function(fn) {
  if (typeof fn !== 'function') {
    throw new TypeError('A setter must be a function.');
  }
  this.setters.push(fn);
  return this;
};



SchemaType.prototype.get = function(fn) {
  if (typeof fn !== 'function') {
    throw new TypeError('A getter must be a function.');
  }
  this.getters.push(fn);
  return this;
};



SchemaType.prototype.validateAll = function(validators) {
  for (let i = 0; i < validators.length; i++) {
    this.validate(validators[i]);
  }
  return this;
};



SchemaType.prototype.validate = function(obj, message, type) {
  if (typeof obj === 'function' || obj && utils.getFunctionName(obj.constructor) === 'RegExp') {
    let properties;
    if (typeof message === 'function') {
      properties = { validator: obj, message: message };
      properties.type = type || 'user defined';
    } else if (message instanceof Object && !type) {
      properties = isSimpleValidator(message) ? Object.assign({}, message) : clone(message);
      if (!properties.message) {
        properties.message = properties.msg;
      }
      properties.validator = obj;
      properties.type = properties.type || 'user defined';
    } else {
      if (message == null) {
        message = MongooseError.messages.general.default;
      }
      if (!type) {
        type = 'user defined';
      }
      properties = { message: message, type: type, validator: obj };
    }

    this.validators.push(properties);
    return this;
  }

  let i;
  let length;
  let arg;

  for (i = 0, length = arguments.length; i < length; i++) {
    arg = arguments[i];
    if (!utils.isPOJO(arg)) {
      const msg = 'Invalid validator. Received (' + typeof arg + ') '
        + arg
        + '. See https://mongoosejs.com/docs/api/schematype.html#SchemaType.prototype.validate()';

      throw new MongooseError(msg);
    }
    this.validate(arg.validator, arg);
  }

  return this;
};



SchemaType.prototype.required = function(required, message) {
  let customOptions = {};

  if (arguments.length > 0 && required == null) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.requiredValidator;
    }, this);

    this.isRequired = false;
    delete this.originalRequiredValue;
    return this;
  }

  if (typeof required === 'object') {
    customOptions = required;
    message = customOptions.message || message;
    required = required.isRequired;
  }

  if (required === false) {
    this.validators = this.validators.filter(function(v) {
      return v.validator !== this.requiredValidator;
    }, this);

    this.isRequired = false;
    delete this.originalRequiredValue;
    return this;
  }

  const _this = this;
  this.isRequired = true;
  this.originalRequiredValue = required;

  if (typeof this.originalRequiredValue !== 'function' &&
      (utils.hasUserDefinedProperty(this.options, 'allowNull') || this.allowNullValidator != null)) {
    throw new MongooseError('Path "' + this.path + '" may not have `allowNull` specified when `required` is true');
  }

  this.requiredValidator = function(v) {
    const cachedRequired = this?.$__?.cachedRequired;

    
    if (cachedRequired != null && !this.$__isSelected(_this.path) && !this[documentIsModified](_this.path)) {
      return true;
    }

    
    
    
    if (cachedRequired != null && _this.path in cachedRequired) {
      const res = cachedRequired[_this.path] ?
        _this.checkRequired(v, this) :
        true;
      delete cachedRequired[_this.path];
      return res;
    } else if (typeof required === 'function') {
      return required.apply(this) ? _this.checkRequired(v, this) : true;
    }

    return _this.checkRequired(v, this);
  };
  if (typeof required === 'string') {
    message = required;
    required = undefined;
  }

  const msg = message || MongooseError.messages.general.required;
  this.validators.unshift(Object.assign({}, customOptions, {
    validator: this.requiredValidator,
    message: msg,
    type: 'required'
  }));

  return this;
};



SchemaType.prototype.allowNull = function(allowNull) {
  if (arguments.length > 0 && this.isRequired && typeof this.originalRequiredValue !== 'function') {
    throw new MongooseError('Path "' + this.path + '" may not have `allowNull` specified when `required` is true');
  }

  this.validators = this.validators.filter(function(v) {
    return v.validator !== this.allowNullValidator;
  }, this);

  if (allowNull !== false) {
    delete this.options.allowNull;
    delete this.allowNullValidator;
    return this;
  }

  this.options.allowNull = false;
  this.allowNullValidator = function(v) {
    return v !== null;
  };

  this.validators.push({
    validator: this.allowNullValidator,
    message: MongooseError.messages.general.allowNull,
    type: 'allowNull'
  });

  return this;
};



SchemaType.prototype.ref = function(ref) {
  this.options.ref = ref;
  return this;
};



SchemaType.prototype.getDefault = function getDefault(parentDoc, init, options) {
  let ret;
  if (this.defaultValue == null) {
    return this.defaultValue;
  }
  if (typeof this.defaultValue === 'function') {
    
    
    
    
    const context = options?.context ?? parentDoc;
    if (
      this.defaultValue === Date.now ||
      this.defaultValue === Array ||
      this.defaultValue.name.toLowerCase() === 'objectid'
    ) {
      ret = this.defaultValue.call(context);
    } else {
      ret = this.defaultValue.call(context, context);
    }
  } else {
    ret = this.defaultValue;
  }

  if (ret != null) {
    if (typeof ret === 'object' && !this.options?.shared) {
      ret = clone(ret);
    }

    if (options?.skipCast) {
      return this._applySetters(ret, parentDoc);
    }

    const casted = this.applySetters(ret, parentDoc, init, undefined, setOptionsForDefaults);
    if (casted && !Array.isArray(casted) && casted.$isSingleNested) {
      casted.$__parent = parentDoc;
    }
    return casted;
  }
  return ret;
};



SchemaType.prototype._applySetters = function(value, scope, init, priorVal, options) {
  let v = value;
  if (init) {
    return v;
  }
  const setters = this.setters;

  for (let i = setters.length - 1; i >= 0; i--) {
    v = setters[i].call(scope, v, priorVal, this, options);
  }

  return v;
};



SchemaType.prototype._castNullish = function _castNullish(v) {
  return v;
};



SchemaType.prototype.applySetters = function(value, scope, init, priorVal, options) {
  let v = this._applySetters(value, scope, init, priorVal, options);
  if (v == null) {
    return this._castNullish(v);
  }
  
  v = this.cast(v, scope, init, priorVal, options);

  return v;
};



SchemaType.prototype.applyGetters = function(value, scope) {
  let v = value;
  const getters = this.getters;
  const len = getters.length;

  if (len === 0) {
    return v;
  }

  for (let i = 0; i < len; ++i) {
    v = getters[i].call(scope, v, this);
  }

  return v;
};



SchemaType.prototype.select = function select(val) {
  this.selected = !!val;
  return this;
};



SchemaType.prototype.doValidate = async function doValidate(value, scope, options) {
  let err = false;
  const path = this.path;

  
  const validators = this.validators.
    filter(v => typeof v === 'object' && v !== null);

  if (!validators.length) {
    return;
  }

  const promises = [];
  for (let i = 0, len = validators.length; i < len; ++i) {
    const v = validators[i];
    const validator = v.validator;
    let ok;

    const validatorProperties = isSimpleValidator(v) ? Object.assign({}, v) : clone(v);
    validatorProperties.path = options?.path || path;
    validatorProperties.fullPath = this.$fullPath;
    validatorProperties.value = value;
    if (typeof value === 'string') {
      validatorProperties.length = value.length;
    }

    if (value === undefined && validator !== this.requiredValidator) {
      continue;
    }
    if (validator instanceof RegExp) {
      ok = validator.test(value);
      if (ok === false) {
        const ErrorConstructor = validatorProperties.ErrorConstructor || ValidatorError;
        err = new ErrorConstructor(validatorProperties, scope);
        err[validatorErrorSymbol] = true;
        throw err;
      }
      continue;
    } else if (typeof validator !== 'function') {
      continue;
    }

    try {
      if (validatorProperties.propsParameter) {
        ok = validator.call(scope, value, validatorProperties);
      } else {
        ok = validator.call(scope, value);
      }
    } catch (error) {
      ok = false;
      validatorProperties.reason = error;
      if (error.message) {
        validatorProperties.message = error.message;
      }
    }

    if (typeof ok?.then === 'function') {
      promises.push(
        ok.then(
          function(ok) {
            if (ok === false) {
              const ErrorConstructor = validatorProperties.ErrorConstructor || ValidatorError;
              err = new ErrorConstructor(validatorProperties, scope);
              err[validatorErrorSymbol] = true;
              throw err;
            }
          },
          function(error) {
            validatorProperties.reason = error;
            validatorProperties.message = error.message;
            ok = false;
            const ErrorConstructor = validatorProperties.ErrorConstructor || ValidatorError;
            err = new ErrorConstructor(validatorProperties, scope);
            err[validatorErrorSymbol] = true;
            throw err;
          })
      );
    } else if (ok !== undefined && !ok) {
      const ErrorConstructor = validatorProperties.ErrorConstructor || ValidatorError;
      err = new ErrorConstructor(validatorProperties, scope);
      err[validatorErrorSymbol] = true;
      throw err;
    }
  }

  await Promise.all(promises);
};


function _validate(ok, validatorProperties) {
  if (ok !== undefined && !ok) {
    const ErrorConstructor = validatorProperties.ErrorConstructor || ValidatorError;
    const err = new ErrorConstructor(validatorProperties);
    err[validatorErrorSymbol] = true;
    return err;
  }
}



SchemaType.prototype.doValidateSync = function(value, scope, options) {
  const path = this.path;
  const count = this.validators.length;

  if (!count) {
    return null;
  }

  let validators = this.validators;
  if (value === void 0) {
    if (this.validators.length !== 0 && this.validators[0].type === 'required') {
      validators = [this.validators[0]];
    } else {
      return null;
    }
  }

  let err = null;
  let i = 0;
  const len = validators.length;
  for (i = 0; i < len; ++i) {
    const v = validators[i];

    if (v === null || typeof v !== 'object') {
      continue;
    }

    const validator = v.validator;
    const validatorProperties = isSimpleValidator(v) ? Object.assign({}, v) : clone(v);
    validatorProperties.path = options?.path || path;
    validatorProperties.fullPath = this.$fullPath;
    validatorProperties.value = value;
    if (typeof value === 'string') {
      validatorProperties.length = value.length;
    }
    let ok = false;

    
    
    if (isAsyncFunction(validator)) {
      continue;
    }

    if (validator instanceof RegExp) {
      err = _validate(validator.test(value), validatorProperties);
      continue;
    }

    if (typeof validator !== 'function') {
      continue;
    }

    try {
      if (validatorProperties.propsParameter) {
        ok = validator.call(scope, value, validatorProperties);
      } else {
        ok = validator.call(scope, value);
      }
    } catch (error) {
      ok = false;
      validatorProperties.reason = error;
    }

    
    
    if (typeof ok?.then === 'function') {
      continue;
    }
    err = _validate(ok, validatorProperties);
    if (err) {
      break;
    }
  }

  return err;
};



SchemaType._isRef = function(self, value, doc, init) {
  
  let ref = init && (self.options?.ref || self.options?.refPath);

  if (!ref && doc?.$__ != null) {
    
    
    
    const path = doc.$__fullPath(self.path, true);

    const owner = doc.ownerDocument();
    ref = (path != null && owner.$populated(path)) || doc.$populated(self.path);
  }

  if (ref) {
    if (value == null) {
      return true;
    }
    if (!Buffer.isBuffer(value) && 
      value._bsontype !== 'Binary' 
      && utils.isObject(value) 
    ) {
      return true;
    }

    return init;
  }

  return false;
};



SchemaType.prototype._castRef = function _castRef(value, doc, init, options) {
  if (value == null) {
    return value;
  }

  if (value.$__ != null) {
    value.$__.wasPopulated = value.$__.wasPopulated || { value: value._doc._id };
    return value;
  }

  
  if (Buffer.isBuffer(value) || !utils.isObject(value)) {
    if (init) {
      return value;
    }
    throw new CastError(this.instance, value, this.path, null, this);
  }

  
  
  
  const path = doc.$__fullPath(this.path, true);
  const owner = doc.ownerDocument();
  const pop = owner.$populated(path, true);

  let ret = value;
  if (!doc.$__.populated ||
    !doc.$__.populated[path] ||
    !doc.$__.populated[path].options ||
    !doc.$__.populated[path].options.options ||
    !doc.$__.populated[path].options.options.lean) {
    const PopulatedModel = pop ? pop.options[populateModelSymbol] : doc.constructor.db.model(this.options.ref);
    ret = PopulatedModel.hydrate(value, null, options);
    ret.$__.wasPopulated = { value: ret._doc._id, options: { [populateModelSymbol]: PopulatedModel } };
  }

  return ret;
};



function handleSingle(val, context) {
  return this.castForQuery(null, val, context);
}



function handleArray(val, context) {
  const _this = this;
  if (!Array.isArray(val)) {
    return [this.castForQuery(null, val, context)];
  }
  return val.map(function(m) {
    return _this.castForQuery(null, m, context);
  });
}



function handle$in(val, context) {
  const _this = this;
  if (!Array.isArray(val)) {
    return [this.castForQuery(null, val, context)];
  }
  return val.map(function(m) {
    if (Array.isArray(m) && m.length === 0) {
      return m;
    }
    return _this.castForQuery(null, m, context);
  });
}



SchemaType.prototype.$conditionalHandlers = {
  $all: handleArray,
  $eq: handleSingle,
  $in: handle$in,
  $ne: handleSingle,
  $nin: handle$in,
  $exists: $exists,
  $type: $type
};



SchemaType.prototype.castForQuery = function($conditional, val, context) {
  let handler;
  if ($conditional != null) {
    handler = this.$conditionalHandlers[$conditional];
    if (!handler) {
      throw new MongooseError('Can\'t use ' + $conditional);
    }
    return handler.call(this, val, context);
  }

  try {
    return this.applySetters(val, context);
  } catch (err) {
    if (err instanceof CastError && err.path === this.path && this.$fullPath != null) {
      err.path = this.$fullPath;
    }
    throw err;
  }
};



SchemaType.checkRequired = function(fn) {
  if (arguments.length !== 0) {
    this._checkRequired = fn;
  }

  return this._checkRequired;
};



SchemaType.prototype.checkRequired = function(val) {
  return val != null;
};



SchemaType.prototype.clone = function() {
  const options = Object.assign({}, this.options);
  const schematype = new this.constructor(this.path, options, this.instance, this.parentSchema);
  schematype.validators = this.validators.slice();
  if (this.requiredValidator !== undefined) schematype.requiredValidator = this.requiredValidator;
  if (this.allowNullValidator !== undefined) schematype.allowNullValidator = this.allowNullValidator;
  if (this.defaultValue !== undefined) schematype.defaultValue = this.defaultValue;
  if (this.$immutable !== undefined && this.options.immutable === undefined) {
    schematype.$immutable = this.$immutable;

    handleImmutable(schematype);
  }
  if (this._index !== undefined) schematype._index = this._index;
  if (this.selected !== undefined) schematype.selected = this.selected;
  if (this.isRequired !== undefined) schematype.isRequired = this.isRequired;
  if (this.originalRequiredValue !== undefined) schematype.originalRequiredValue = this.originalRequiredValue;
  schematype.getters = this.getters.slice();
  schematype.setters = this.setters.slice();
  return schematype;
};



SchemaType.prototype.getEmbeddedSchemaType = function getEmbeddedSchemaType() {
  return this.embeddedSchemaType;
};



SchemaType.prototype._duplicateKeyErrorMessage = null;



SchemaType.prototype.toJSONSchema = function toJSONSchema(_options) { 
  throw new MongooseError(`Converting unsupported SchemaType to JSON Schema: ${this.instance} at path "${this.path}"`);
};


SchemaType.prototype.autoEncryptionType = function autoEncryptionType() {
  return null;
};



module.exports = exports = SchemaType;

exports.CastError = CastError;

exports.ValidatorError = ValidatorError;
