// mongodb/lib/deps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getKerberos = getKerberos;
exports.getZstdLibrary = getZstdLibrary;
exports.getAwsCredentialProvider = getAwsCredentialProvider;
exports.getGcpMetadata = getGcpMetadata;
exports.getSnappy = getSnappy;
exports.getSocks = getSocks;
exports.getMongoDBClientEncryption = getMongoDBClientEncryption;
const error_1 = require("./error");
function makeErrorModule(error) {
    const props = error ? { kModuleError: error } : {};
    return new Proxy(props, {
        get: (_, key) => {
            if (key === 'kModuleError') {
                return error;
            }
            throw error;
        },
        set: () => {
            throw error;
        }
    });
}
function getKerberos() {
    let kerberos;
    try {
        
        
        kerberos = require('kerberos');
    }
    catch (error) {
        kerberos = makeErrorModule(new error_1.MongoMissingDependencyError('Optional module `kerberos` not found. Please install it to enable kerberos authentication', { cause: error, dependencyName: 'kerberos' }));
    }
    return kerberos;
}
function getZstdLibrary() {
    let ZStandard;
    try {
        
        ZStandard = require('@mongodb-js/zstd');
    }
    catch (error) {
        ZStandard = makeErrorModule(new error_1.MongoMissingDependencyError('Optional module `@mongodb-js/zstd` not found. Please install it to enable zstd compression', { cause: error, dependencyName: 'zstd' }));
    }
    return ZStandard;
}
function getAwsCredentialProvider() {
    try {
        
        
        const credentialProvider = require('@aws-sdk/credential-providers');
        return credentialProvider;
    }
    catch (error) {
        return makeErrorModule(new error_1.MongoMissingDependencyError('Optional module `@aws-sdk/credential-providers` not found.' +
            ' Please install it to enable getting aws credentials via the official sdk.', { cause: error, dependencyName: '@aws-sdk/credential-providers' }));
    }
}
function getGcpMetadata() {
    try {
        
        
        const credentialProvider = require('gcp-metadata');
        return credentialProvider;
    }
    catch (error) {
        return makeErrorModule(new error_1.MongoMissingDependencyError('Optional module `gcp-metadata` not found.' +
            ' Please install it to enable getting gcp credentials via the official sdk.', { cause: error, dependencyName: 'gcp-metadata' }));
    }
}
function getSnappy() {
    try {
        
        
        const value = require('snappy');
        return value;
    }
    catch (error) {
        const kModuleError = new error_1.MongoMissingDependencyError('Optional module `snappy` not found. Please install it to enable snappy compression', { cause: error, dependencyName: 'snappy' });
        return { kModuleError };
    }
}
function getSocks() {
    try {
        
        
        const value = require('socks');
        return value;
    }
    catch (error) {
        const kModuleError = new error_1.MongoMissingDependencyError('Optional module `socks` not found. Please install it to connections over a SOCKS5 proxy', { cause: error, dependencyName: 'socks' });
        return { kModuleError };
    }
}

function getMongoDBClientEncryption() {
    let mongodbClientEncryption = null;
    try {
        
        
        
        
        mongodbClientEncryption = require('mongodb-client-encryption');
    }
    catch (error) {
        const kModuleError = new error_1.MongoMissingDependencyError('Optional module `mongodb-client-encryption` not found. Please install it to use auto encryption or ClientEncryption.', { cause: error, dependencyName: 'mongodb-client-encryption' });
        return { kModuleError };
    }
    return mongodbClientEncryption;
}
