// mongodb/lib/sdam/server_selection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeprioritizedServers = exports.MIN_SECONDARY_WRITE_WIRE_VERSION = void 0;
exports.writableServerSelector = writableServerSelector;
exports.sameServerSelector = sameServerSelector;
exports.secondaryWritableServerSelector = secondaryWritableServerSelector;
exports.readPreferenceServerSelector = readPreferenceServerSelector;
const error_1 = require("../error");
const read_preference_1 = require("../read_preference");
const common_1 = require("./common");

const IDLE_WRITE_PERIOD = 10000;
const SMALLEST_MAX_STALENESS_SECONDS = 90;

exports.MIN_SECONDARY_WRITE_WIRE_VERSION = 13;

class DeprioritizedServers {
    constructor(descriptions) {
        this.deprioritized = new Set();
        for (const description of descriptions ?? []) {
            this.add(description);
        }
    }
    add({ address }) {
        this.deprioritized.add(address);
    }
    has({ address }) {
        return this.deprioritized.has(address);
    }
}
exports.DeprioritizedServers = DeprioritizedServers;
function filterDeprioritized(candidates, deprioritized) {
    const filtered = candidates.filter(candidate => !deprioritized.has(candidate));
    return filtered.length ? filtered : candidates;
}

function writableServerSelector() {
    return function writableServer(topologyDescription, servers, deprioritized) {
        const eligibleServers = filterDeprioritized(servers.filter(({ isWritable }) => isWritable), deprioritized);
        return latencyWindowReducer(topologyDescription, eligibleServers);
    };
}

function sameServerSelector(description) {
    return function sameServerSelector(_topologyDescription, servers, _deprioritized) {
        if (!description)
            return [];
        
        
        return servers.filter(sd => {
            return sd.address === description.address && sd.type !== common_1.ServerType.Unknown;
        });
    };
}

function secondaryWritableServerSelector(wireVersion, readPreference) {
    
    
    
    
    if (!readPreference ||
        !wireVersion ||
        (wireVersion && wireVersion < exports.MIN_SECONDARY_WRITE_WIRE_VERSION)) {
        return readPreferenceServerSelector(read_preference_1.ReadPreference.primary);
    }
    return readPreferenceServerSelector(readPreference);
}

function maxStalenessReducer(readPreference, topologyDescription, servers) {
    if (readPreference.maxStalenessSeconds == null || readPreference.maxStalenessSeconds < 0) {
        return servers;
    }
    const maxStaleness = readPreference.maxStalenessSeconds;
    const maxStalenessVariance = (topologyDescription.heartbeatFrequencyMS + IDLE_WRITE_PERIOD) / 1000;
    if (maxStaleness < maxStalenessVariance) {
        throw new error_1.MongoInvalidArgumentError(`Option "maxStalenessSeconds" must be at least ${maxStalenessVariance} seconds`);
    }
    if (maxStaleness < SMALLEST_MAX_STALENESS_SECONDS) {
        throw new error_1.MongoInvalidArgumentError(`Option "maxStalenessSeconds" must be at least ${SMALLEST_MAX_STALENESS_SECONDS} seconds`);
    }
    if (topologyDescription.type === common_1.TopologyType.ReplicaSetWithPrimary) {
        const primary = Array.from(topologyDescription.servers.values()).filter(primaryFilter)[0];
        return servers.filter((server) => {
            const stalenessMS = server.lastUpdateTime -
                server.lastWriteDate -
                (primary.lastUpdateTime - primary.lastWriteDate) +
                topologyDescription.heartbeatFrequencyMS;
            const staleness = stalenessMS / 1000;
            const maxStalenessSeconds = readPreference.maxStalenessSeconds ?? 0;
            return staleness <= maxStalenessSeconds;
        });
    }
    if (topologyDescription.type === common_1.TopologyType.ReplicaSetNoPrimary) {
        if (servers.length === 0) {
            return servers;
        }
        const sMax = servers.reduce((max, s) => s.lastWriteDate > max.lastWriteDate ? s : max);
        return servers.filter((server) => {
            const stalenessMS = sMax.lastWriteDate - server.lastWriteDate + topologyDescription.heartbeatFrequencyMS;
            const staleness = stalenessMS / 1000;
            const maxStalenessSeconds = readPreference.maxStalenessSeconds ?? 0;
            return staleness <= maxStalenessSeconds;
        });
    }
    return servers;
}

function tagSetMatch(tagSet, serverTags) {
    return Object.entries(tagSet).every(([key, value]) => serverTags[key] != null && serverTags[key] === value);
}

function tagSetReducer({ tags }, servers) {
    if (tags == null || tags.length === 0) {
        
        return servers;
    }
    for (const tagSet of tags) {
        const serversMatchingTagset = servers.filter((s) => tagSetMatch(tagSet, s.tags));
        if (serversMatchingTagset.length) {
            return serversMatchingTagset;
        }
    }
    return [];
}

function latencyWindowReducer(topologyDescription, servers) {
    const low = servers.reduce((min, server) => Math.min(server.roundTripTime, min), Infinity);
    const high = low + topologyDescription.localThresholdMS;
    return servers.filter(server => server.roundTripTime <= high && server.roundTripTime >= low);
}

function primaryFilter(server) {
    return server.type === common_1.ServerType.RSPrimary;
}
function secondaryFilter(server) {
    return server.type === common_1.ServerType.RSSecondary;
}
function nearestFilter(server) {
    return server.type === common_1.ServerType.RSSecondary || server.type === common_1.ServerType.RSPrimary;
}
function knownFilter(server) {
    return server.type !== common_1.ServerType.Unknown;
}
function loadBalancerFilter(server) {
    return server.type === common_1.ServerType.LoadBalancer;
}
function isDeprioritizedFactory(deprioritized) {
    return server => 
    
    !deprioritized.has(server);
}
function secondarySelector(readPreference, topologyDescription, servers, deprioritized) {
    const mode = readPreference.mode;
    switch (mode) {
        case 'primary':
            
            
            
            
            return servers.filter(primaryFilter);
        case 'primaryPreferred': {
            const primary = servers.filter(primaryFilter);
            
            
            const eligiblePrimary = primary.filter(isDeprioritizedFactory(deprioritized));
            if (eligiblePrimary.length) {
                return eligiblePrimary;
            }
            
            
            
            
            const secondaries = tagSetReducer(readPreference, maxStalenessReducer(readPreference, topologyDescription, servers.filter(secondaryFilter)));
            const eligibleSecondaries = secondaries.filter(isDeprioritizedFactory(deprioritized));
            if (eligibleSecondaries.length) {
                return latencyWindowReducer(topologyDescription, eligibleSecondaries);
            }
            
            
            
            return primary.length ? primary : latencyWindowReducer(topologyDescription, secondaries);
        }
        case 'nearest': {
            const eligible = filterDeprioritized(tagSetReducer(readPreference, maxStalenessReducer(readPreference, topologyDescription, servers.filter(nearestFilter))), deprioritized);
            return latencyWindowReducer(topologyDescription, eligible);
        }
        case 'secondary':
        case 'secondaryPreferred': {
            const secondaries = tagSetReducer(readPreference, maxStalenessReducer(readPreference, topologyDescription, servers.filter(secondaryFilter)));
            const eligibleSecondaries = secondaries.filter(isDeprioritizedFactory(deprioritized));
            if (eligibleSecondaries.length) {
                return latencyWindowReducer(topologyDescription, eligibleSecondaries);
            }
            
            if (mode === read_preference_1.ReadPreference.SECONDARY_PREFERRED) {
                const primary = servers.filter(primaryFilter);
                
                
                const eligiblePrimary = primary.filter(isDeprioritizedFactory(deprioritized));
                if (eligiblePrimary.length)
                    return eligiblePrimary;
                
                return secondaries.length
                    ? latencyWindowReducer(topologyDescription, secondaries)
                    : primary;
            }
            
            return latencyWindowReducer(topologyDescription, secondaries);
        }
        default: {
            const _exhaustiveCheck = mode;
            throw new error_1.MongoRuntimeError(`unexpected readPreference=${mode} (should never happen).  Please report a bug in the Node driver Jira project.`);
        }
    }
}

function readPreferenceServerSelector(readPreference) {
    if (!readPreference.isValid()) {
        throw new error_1.MongoInvalidArgumentError('Invalid read preference specified');
    }
    return function readPreferenceServers(topologyDescription, servers, deprioritized) {
        switch (topologyDescription.type) {
            case 'Single':
                return latencyWindowReducer(topologyDescription, servers.filter(knownFilter));
            case 'ReplicaSetNoPrimary':
            case 'ReplicaSetWithPrimary':
                return secondarySelector(readPreference, topologyDescription, servers, deprioritized);
            case 'Sharded': {
                const selectable = filterDeprioritized(servers, deprioritized);
                return latencyWindowReducer(topologyDescription, selectable.filter(knownFilter));
            }
            case 'Unknown':
                return [];
            case 'LoadBalanced':
                return servers.filter(loadBalancerFilter);
            default: {
                const _exhaustiveCheck = topologyDescription.type;
                throw new error_1.MongoRuntimeError(`unexpected topology type: ${topologyDescription.type} (this should never happen).  Please file a bug in the Node driver Jira project.`);
            }
        }
    };
}
