// mongodb/lib/cmap/connection_pool_events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionPoolClearedEvent = exports.ConnectionCheckedInEvent = exports.ConnectionCheckedOutEvent = exports.ConnectionCheckOutFailedEvent = exports.ConnectionCheckOutStartedEvent = exports.ConnectionClosedEvent = exports.ConnectionReadyEvent = exports.ConnectionCreatedEvent = exports.ConnectionPoolClosedEvent = exports.ConnectionPoolReadyEvent = exports.ConnectionPoolCreatedEvent = exports.ConnectionPoolMonitoringEvent = void 0;
const constants_1 = require("../constants");
const utils_1 = require("../utils");

class ConnectionPoolMonitoringEvent {
    
    constructor(pool) {
        this.time = new Date();
        this.address = pool.address;
    }
}
exports.ConnectionPoolMonitoringEvent = ConnectionPoolMonitoringEvent;

class ConnectionPoolCreatedEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool) {
        super(pool);
        
        this.name = constants_1.CONNECTION_POOL_CREATED;
        const { maxConnecting, maxPoolSize, minPoolSize, maxIdleTimeMS, waitQueueTimeoutMS } = pool.options;
        this.options = { maxConnecting, maxPoolSize, minPoolSize, maxIdleTimeMS, waitQueueTimeoutMS };
    }
}
exports.ConnectionPoolCreatedEvent = ConnectionPoolCreatedEvent;

class ConnectionPoolReadyEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool) {
        super(pool);
        
        this.name = constants_1.CONNECTION_POOL_READY;
    }
}
exports.ConnectionPoolReadyEvent = ConnectionPoolReadyEvent;

class ConnectionPoolClosedEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool) {
        super(pool);
        
        this.name = constants_1.CONNECTION_POOL_CLOSED;
    }
}
exports.ConnectionPoolClosedEvent = ConnectionPoolClosedEvent;

class ConnectionCreatedEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool, connection) {
        super(pool);
        
        this.name = constants_1.CONNECTION_CREATED;
        this.connectionId = connection.id;
    }
}
exports.ConnectionCreatedEvent = ConnectionCreatedEvent;

class ConnectionReadyEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool, connection, connectionCreatedEventTime) {
        super(pool);
        
        this.name = constants_1.CONNECTION_READY;
        this.durationMS = (0, utils_1.processTimeMS)() - connectionCreatedEventTime;
        this.connectionId = connection.id;
    }
}
exports.ConnectionReadyEvent = ConnectionReadyEvent;

class ConnectionClosedEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool, connection, reason, error) {
        super(pool);
        
        this.name = constants_1.CONNECTION_CLOSED;
        this.connectionId = connection.id;
        this.reason = reason;
        this.serviceId = connection.serviceId;
        this.error = error ?? null;
    }
}
exports.ConnectionClosedEvent = ConnectionClosedEvent;

class ConnectionCheckOutStartedEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool) {
        super(pool);
        
        this.name = constants_1.CONNECTION_CHECK_OUT_STARTED;
    }
}
exports.ConnectionCheckOutStartedEvent = ConnectionCheckOutStartedEvent;

class ConnectionCheckOutFailedEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool, reason, checkoutTime, error) {
        super(pool);
        
        this.name = constants_1.CONNECTION_CHECK_OUT_FAILED;
        this.durationMS = (0, utils_1.processTimeMS)() - checkoutTime;
        this.reason = reason;
        this.error = error;
    }
}
exports.ConnectionCheckOutFailedEvent = ConnectionCheckOutFailedEvent;

class ConnectionCheckedOutEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool, connection, checkoutTime) {
        super(pool);
        
        this.name = constants_1.CONNECTION_CHECKED_OUT;
        this.durationMS = (0, utils_1.processTimeMS)() - checkoutTime;
        this.connectionId = connection.id;
    }
}
exports.ConnectionCheckedOutEvent = ConnectionCheckedOutEvent;

class ConnectionCheckedInEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool, connection) {
        super(pool);
        
        this.name = constants_1.CONNECTION_CHECKED_IN;
        this.connectionId = connection.id;
    }
}
exports.ConnectionCheckedInEvent = ConnectionCheckedInEvent;

class ConnectionPoolClearedEvent extends ConnectionPoolMonitoringEvent {
    
    constructor(pool, options = {}) {
        super(pool);
        
        this.name = constants_1.CONNECTION_POOL_CLEARED;
        this.serviceId = options.serviceId;
        this.interruptInUseConnections = options.interruptInUseConnections;
    }
}
exports.ConnectionPoolClearedEvent = ConnectionPoolClearedEvent;
