// mongodb/lib/cmap/auth/aws_temporary_credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AWSSDKCredentialProvider = void 0;
const process = require("process");
const deps_1 = require("../../deps");
const error_1 = require("../../error");

class AWSSDKCredentialProvider {
    
    constructor(credentialsProvider) {
        if (credentialsProvider) {
            this._provider = credentialsProvider;
        }
    }
    static get awsSDK() {
        AWSSDKCredentialProvider._awsSDK ??= (0, deps_1.getAwsCredentialProvider)();
        return AWSSDKCredentialProvider._awsSDK;
    }
    
    get provider() {
        if ('kModuleError' in AWSSDKCredentialProvider.awsSDK) {
            throw AWSSDKCredentialProvider.awsSDK.kModuleError;
        }
        if (this._provider) {
            return this._provider;
        }
        let { AWS_STS_REGIONAL_ENDPOINTS = '', AWS_REGION = '' } = process.env;
        AWS_STS_REGIONAL_ENDPOINTS = AWS_STS_REGIONAL_ENDPOINTS.toLowerCase();
        AWS_REGION = AWS_REGION.toLowerCase();
        
        const awsRegionSettingsExist = AWS_REGION.length !== 0 && AWS_STS_REGIONAL_ENDPOINTS.length !== 0;
        
        const LEGACY_REGIONS = new Set([
            'ap-northeast-1',
            'ap-south-1',
            'ap-southeast-1',
            'ap-southeast-2',
            'aws-global',
            'ca-central-1',
            'eu-central-1',
            'eu-north-1',
            'eu-west-1',
            'eu-west-2',
            'eu-west-3',
            'sa-east-1',
            'us-east-1',
            'us-east-2',
            'us-west-1',
            'us-west-2'
        ]);
        
        const useRegionalSts = AWS_STS_REGIONAL_ENDPOINTS === 'regional' ||
            (AWS_STS_REGIONAL_ENDPOINTS === 'legacy' && !LEGACY_REGIONS.has(AWS_REGION));
        this._provider =
            awsRegionSettingsExist && useRegionalSts
                ? AWSSDKCredentialProvider.awsSDK.fromNodeProviderChain({
                    clientConfig: { region: AWS_REGION }
                })
                : AWSSDKCredentialProvider.awsSDK.fromNodeProviderChain();
        return this._provider;
    }
    async getCredentials() {
        
        try {
            const creds = await this.provider();
            return {
                AccessKeyId: creds.accessKeyId,
                SecretAccessKey: creds.secretAccessKey,
                Token: creds.sessionToken,
                Expiration: creds.expiration
            };
        }
        catch (error) {
            throw new error_1.MongoAWSError(error.message, { cause: error });
        }
    }
}
exports.AWSSDKCredentialProvider = AWSSDKCredentialProvider;
