// mongoose/lib/schema/bigint.js
'use strict';



const CastError = require('../error/cast');
const SchemaType = require('../schemaType');
const castBigInt = require('../cast/bigint');



function SchemaBigInt(path, options, _schemaOptions, parentSchema) {
  SchemaType.call(this, path, options, 'BigInt', parentSchema);
}


SchemaBigInt.schemaName = 'BigInt';

SchemaBigInt.defaultOptions = {};


SchemaBigInt.prototype = Object.create(SchemaType.prototype);
SchemaBigInt.prototype.constructor = SchemaBigInt;



SchemaBigInt._cast = castBigInt;



SchemaBigInt.set = SchemaType.set;

SchemaBigInt.setters = [];



SchemaBigInt.get = SchemaType.get;



SchemaBigInt.cast = function cast(caster) {
  if (arguments.length === 0) {
    return this._cast;
  }
  if (caster === false) {
    caster = this._defaultCaster;
  }
  this._cast = caster;

  return this._cast;
};



SchemaBigInt._checkRequired = v => v != null;



SchemaBigInt.checkRequired = SchemaType.checkRequired;



SchemaBigInt.prototype.checkRequired = function(value) {
  return this.constructor._checkRequired(value);
};



SchemaBigInt.prototype.cast = function(value) {
  let castBigInt;
  if (typeof this._castFunction === 'function') {
    castBigInt = this._castFunction;
  } else if (typeof this.constructor.cast === 'function') {
    castBigInt = this.constructor.cast();
  } else {
    castBigInt = SchemaBigInt.cast();
  }

  try {
    return castBigInt(value);
  } catch (error) {
    throw new CastError('BigInt', value, this.path, error, this);
  }
};



const $conditionalHandlers = {
  ...SchemaType.prototype.$conditionalHandlers,
  $gt: handleSingle,
  $gte: handleSingle,
  $lt: handleSingle,
  $lte: handleSingle
};



Object.defineProperty(SchemaBigInt.prototype, '$conditionalHandlers', {
  enumerable: false,
  value: $conditionalHandlers
});



function handleSingle(val, context) {
  return this.castForQuery(null, val, context);
}



SchemaBigInt.prototype.castForQuery = function($conditional, val, context) {
  let handler;
  if ($conditional != null) {
    handler = this.$conditionalHandlers[$conditional];

    if (handler) {
      return handler.call(this, val);
    }

    return this.applySetters(val, context);
  }

  try {
    return this.applySetters(val, context);
  } catch (err) {
    if (err instanceof CastError && err.path === this.path && this.$fullPath != null) {
      err.path = this.$fullPath;
    }
    throw err;
  }
};



SchemaBigInt.prototype._castNullish = function _castNullish(v) {
  if (typeof v === 'undefined') {
    return v;
  }
  const castBigInt = typeof this.constructor.cast === 'function' ?
    this.constructor.cast() :
    SchemaBigInt.cast();
  if (castBigInt == null) {
    return v;
  }
  return v;
};



SchemaBigInt.prototype.toJSONSchema = function toJSONSchema(options) {
  return this._createJSONSchemaTypeDefinition('string', 'long', options);
};

SchemaBigInt.prototype.autoEncryptionType = function autoEncryptionType() {
  return 'long';
};



module.exports = SchemaBigInt;
