// mongoose/lib/schema/boolean.js
'use strict';



const CastError = require('../error/cast');
const SchemaType = require('../schemaType');
const castBoolean = require('../cast/boolean');



function SchemaBoolean(path, options, _schemaOptions, parentSchema) {
  SchemaType.call(this, path, options, 'Boolean', parentSchema);
}


SchemaBoolean.schemaName = 'Boolean';

SchemaBoolean.defaultOptions = {};


SchemaBoolean.prototype = Object.create(SchemaType.prototype);
SchemaBoolean.prototype.constructor = SchemaBoolean;



SchemaBoolean._cast = castBoolean;



SchemaBoolean.set = SchemaType.set;

SchemaBoolean.setters = [];



SchemaBoolean.get = SchemaType.get;



SchemaBoolean.cast = function cast(caster) {
  if (arguments.length === 0) {
    return this._cast;
  }
  if (caster === false) {
    caster = this._defaultCaster;
  }
  this._cast = caster;

  return this._cast;
};



SchemaBoolean._defaultCaster = v => {
  if (v != null && typeof v !== 'boolean') {
    throw new Error();
  }
  return v;
};



SchemaBoolean._checkRequired = v => v === true || v === false;



SchemaBoolean.checkRequired = SchemaType.checkRequired;



SchemaBoolean.prototype.checkRequired = function(value) {
  return this.constructor._checkRequired(value);
};



Object.defineProperty(SchemaBoolean, 'convertToTrue', {
  get: () => castBoolean.convertToTrue,
  set: v => { castBoolean.convertToTrue = v; }
});



Object.defineProperty(SchemaBoolean, 'convertToFalse', {
  get: () => castBoolean.convertToFalse,
  set: v => { castBoolean.convertToFalse = v; }
});



SchemaBoolean.prototype.cast = function(value) {
  let castBoolean;
  if (typeof this._castFunction === 'function') {
    castBoolean = this._castFunction;
  } else if (typeof this.constructor.cast === 'function') {
    castBoolean = this.constructor.cast();
  } else {
    castBoolean = SchemaBoolean.cast();
  }

  try {
    return castBoolean(value);
  } catch (error) {
    throw new CastError('Boolean', value, this.path, error, this);
  }
};

const $conditionalHandlers = { ...SchemaType.prototype.$conditionalHandlers };



Object.defineProperty(SchemaBoolean.prototype, '$conditionalHandlers', {
  enumerable: false,
  value: $conditionalHandlers
});



SchemaBoolean.prototype.castForQuery = function($conditional, val, context) {
  let handler;
  if ($conditional != null) {
    handler = this.$conditionalHandlers[$conditional];

    if (handler) {
      return handler.call(this, val);
    }

    return this.applySetters(val, context);
  }

  try {
    return this.applySetters(val, context);
  } catch (err) {
    if (err instanceof CastError && err.path === this.path && this.$fullPath != null) {
      err.path = this.$fullPath;
    }
    throw err;
  }
};



SchemaBoolean.prototype._castNullish = function _castNullish(v) {
  if (typeof v === 'undefined') {
    return v;
  }
  const castBoolean = typeof this.constructor.cast === 'function' ?
    this.constructor.cast() :
    SchemaBoolean.cast();
  if (castBoolean == null) {
    return v;
  }
  if (castBoolean.convertToFalse instanceof Set && castBoolean.convertToFalse.has(v)) {
    return false;
  }
  if (castBoolean.convertToTrue instanceof Set && castBoolean.convertToTrue.has(v)) {
    return true;
  }
  return v;
};



SchemaBoolean.prototype.toJSONSchema = function toJSONSchema(options) {
  return this._createJSONSchemaTypeDefinition('boolean', 'bool', options);
};

SchemaBoolean.prototype.autoEncryptionType = function autoEncryptionType() {
  return 'bool';
};



module.exports = SchemaBoolean;
