// mongoose/lib/schema/double.js
'use strict';



const CastError = require('../error/cast');
const SchemaType = require('../schemaType');
const castDouble = require('../cast/double');



function SchemaDouble(path, options, _schemaOptions, parentSchema) {
  SchemaType.call(this, path, options, 'Double', parentSchema);
}


SchemaDouble.schemaName = 'Double';

SchemaDouble.defaultOptions = {};


SchemaDouble.prototype = Object.create(SchemaType.prototype);
SchemaDouble.prototype.constructor = SchemaDouble;



SchemaDouble._cast = castDouble;



SchemaDouble.set = SchemaType.set;

SchemaDouble.setters = [];



SchemaDouble.get = SchemaType.get;



SchemaDouble._defaultCaster = v => {
  if (v != null) {
    if (v._bsontype !== 'Double') {
      throw new Error();
    }
  }

  return v;
};



SchemaDouble.cast = function cast(caster) {
  if (arguments.length === 0) {
    return this._cast;
  }
  if (caster === false) {
    caster = this._defaultCaster;
  }

  this._cast = caster;

  return this._cast;
};




SchemaDouble._checkRequired = v => v != null;


SchemaDouble.checkRequired = SchemaType.checkRequired;



SchemaDouble.prototype.checkRequired = function(value) {
  return this.constructor._checkRequired(value);
};



SchemaDouble.prototype.cast = function(value) {
  let castDouble;
  if (typeof this._castFunction === 'function') {
    castDouble = this._castFunction;
  } else if (typeof this.constructor.cast === 'function') {
    castDouble = this.constructor.cast();
  } else {
    castDouble = SchemaDouble.cast();
  }

  try {
    return castDouble(value);
  } catch (error) {
    throw new CastError('Double', value, this.path, error, this);
  }
};



function handleSingle(val) {
  return this.cast(val);
}

const $conditionalHandlers = {
  ...SchemaType.prototype.$conditionalHandlers,
  $gt: handleSingle,
  $gte: handleSingle,
  $lt: handleSingle,
  $lte: handleSingle
};



Object.defineProperty(SchemaDouble.prototype, '$conditionalHandlers', {
  enumerable: false,
  value: $conditionalHandlers
});



SchemaDouble.prototype.toJSONSchema = function toJSONSchema(options) {
  return this._createJSONSchemaTypeDefinition('number', 'double', options);
};

SchemaDouble.prototype.autoEncryptionType = function autoEncryptionType() {
  return 'double';
};



module.exports = SchemaDouble;
