// mongoose/lib/schema/union.js
'use strict';



const SchemaUnionOptions = require('../options/schemaUnionOptions');
const SchemaType = require('../schemaType');

const firstValueSymbol = Symbol('firstValue');



class Union extends SchemaType {
  
  constructor(key, options, schemaOptions, parentSchema) {
    super(key, options, 'Union', parentSchema);
    if (!Array.isArray(options?.of) || options.of.length === 0) {
      throw new Error('Union schema type requires an array of types');
    }
    this.schemaTypes = options.of.map(obj => parentSchema.interpretAsType(key, obj, schemaOptions));
    this.$isSchemaUnion = true;
  }

  cast(val, doc, init, prev, options) {
    let firstValue = firstValueSymbol;
    let lastError;
    
    
    
    
    
    for (let i = 0; i < this.schemaTypes.length; ++i) {
      try {
        const casted = this.schemaTypes[i].cast(val, doc, init, prev, options);
        if (casted === val) {
          return casted;
        }
        if (firstValue === firstValueSymbol) {
          firstValue = casted;
        }
      } catch (error) {
        lastError = error;
      }
    }
    if (firstValue !== firstValueSymbol) {
      return firstValue;
    }
    throw lastError;
  }

  
  applySetters(val, doc, init, prev, options) {
    let firstValue = firstValueSymbol;
    let lastError;
    
    
    
    
    
    for (let i = 0; i < this.schemaTypes.length; ++i) {
      try {
        let castedVal = this.schemaTypes[i]._applySetters(val, doc, init, prev, options);
        if (castedVal == null) {
          castedVal = this.schemaTypes[i]._castNullish(castedVal);
        } else {
          castedVal = this.schemaTypes[i].cast(castedVal, doc, init, prev, options);
        }
        if (castedVal === val) {
          return castedVal;
        }
        if (firstValue === firstValueSymbol) {
          firstValue = castedVal;
        }
      } catch (error) {
        lastError = error;
      }
    }
    if (firstValue !== firstValueSymbol) {
      return firstValue;
    }
    throw lastError;
  }

  async doValidate(value, scope, options) {
    if (options && options.skipSchemaValidators) {
      if (value != null && typeof value.validate === 'function') {
        return value.validate();
      }
      return;
    }

    await super.doValidate(value, scope, options);
    if (value != null && typeof value.validate === 'function') {
      await value.validate();
    }
  }

  doValidateSync(value, scope, options) {
    if (!options || !options.skipSchemaValidators) {
      const schemaTypeError = super.doValidateSync(value, scope, options);
      if (schemaTypeError) {
        return schemaTypeError;
      }
    }
    if (value != null && typeof value.validateSync === 'function') {
      return value.validateSync();
    }
  }

  clone() {
    const schematype = super.clone();

    schematype.schemaTypes = this.schemaTypes.map(schemaType => schemaType.clone());
    return schematype;
  }

  
  toJSONSchema(options) {
    const isRequired = this.options.required && typeof this.options.required !== 'function';
    const childOptions = { ...options, _overrideRequired: true };
    const jsonSchemas = this.schemaTypes.map(schemaType => schemaType.toJSONSchema(childOptions));
    if (isRequired) {
      return { anyOf: jsonSchemas };
    }

    return {
      anyOf: [
        options?.useBsonType ? { bsonType: 'null' } : { type: 'null' },
        ...jsonSchemas
      ]
    };
  }
}


Union.schemaName = 'Union';

Union.defaultOptions = {};

Union.prototype.OptionsConstructor = SchemaUnionOptions;

module.exports = Union;
